File: Raven.Client.Java/lib/src/main/java/raven/client/document/batches/LazyLoadOperation.java
package raven.client.document.batches;

import java.util.HashMap;
import java.util.Map;

import org.apache.commons.lang.StringUtils;
import org.apache.http.HttpStatus;

import raven.abstractions.data.GetRequest;
import raven.abstractions.data.GetResponse;
import raven.abstractions.data.JsonDocument;
import raven.client.connection.IDatabaseCommands;
import raven.client.connection.SerializationHelper;
import raven.client.document.sessionoperations.LoadOperation;
import raven.client.utils.UrlUtils;

public class LazyLoadOperation<T> implements ILazyOperation {

  private final String key;
  private final LoadOperation loadOperation;
  private final String transformer;
  private Object result;
  private boolean requiresRetry;
  private Class<T> clazz;
  public LazyLoadOperation(Class<T> clazz, String key, LoadOperation loadOperation) {
    this (clazz, key, loadOperation, null);
  }

  public LazyLoadOperation(Class<T> clazz, String key, LoadOperation loadOperation, String transformer) {
    this.key = key;
    this.loadOperation = loadOperation;
    this.clazz = clazz;
    this.transformer = transformer;
  }

  @Override
  public GetRequest createRequest() {
    String path = null;
    if (StringUtils.isNotEmpty(transformer)) {
      path = "/queries/"     + UrlUtils.escapeDataString(key);
      if (StringUtils.isNotEmpty(transformer)) {
        path += "&transformer=" + transformer;
      }
    } else {
      path = "/docs/" + UrlUtils.escapeDataString(key);
    }

    GetRequest request = new GetRequest();
    request.setUrl(path);
    return request;
  }

  @Override
  public Object getResult() {
    return result;
  }

  public void setResult(Object result) {
    this.result = result;
  }

  @Override
  public boolean isRequiresRetry() {
    return requiresRetry;
  }

  public void setRequiresRetry(boolean requiresRetry) {
    this.requiresRetry = requiresRetry;
  }

  @Override
  public void handleResponse(GetResponse response) {
    if(response.getStatus() == HttpStatus.SC_NOT_FOUND) {
      result = null;
      requiresRetry = false;
      return;
    }

    Map<String, String> headers = new HashMap<>(response.getHeaders());
    JsonDocument jsonDocument = SerializationHelper.deserializeJsonDocument(key, response.getResult(), headers, response.getStatus());
    handleResponse(jsonDocument);
  }

  private void handleResponse(JsonDocument jsonDocument) {
    requiresRetry = loadOperation.setResult(jsonDocument);
    if (requiresRetry == false) {
      result = loadOperation.complete(clazz);
    }
  }

  @Override
  public AutoCloseable enterContext() {
    return loadOperation.enterLoadContext();
  }

  public Object executeEmbedded(IDatabaseCommands commands) {
    return commands.get(key);
  }

  public void handleEmbeddedResponse(Object result) {
    handleResponse((JsonDocument) result);
  }

}

File: src/Raven.NewClient/Commands/Lazy/LazyLoadOperation.cs
using System;
using System.Text;

using Raven.NewClient.Abstractions.Data;
using Raven.NewClient.Client.Data;
using Raven.NewClient.Client.Data.Queries;
using Raven.NewClient.Client.Document.Batches;
using Raven.NewClient.Client.Json;
using Sparrow.Json;

namespace Raven.NewClient.Client.Commands.Lazy
{
    public class LazyLoadOperation<T> : ILazyOperation
    {
        private readonly LoadOperation _loadOperation;
        private readonly string[] _ids;
        private readonly string _transformer;
        private readonly string[] _includes;

        public LazyLoadOperation(
            LoadOperation loadOperation,
            string[] ids,
            string[] includes,
            string transformer = null)
        {
            _loadOperation = loadOperation;
            _ids = ids;
            _includes = includes;
            _transformer = transformer;
        }

        public LazyLoadOperation(
            LoadOperation loadOperation,
            string id)
        {
            this._loadOperation = loadOperation;
            _ids = new[] { id };
        }

        public GetRequest CreateRequest()
        {
            var queryBuilder = new StringBuilder("?");
            _includes.ApplyIfNotNull(include => queryBuilder.AppendFormat("&include={0}", include));
            _ids.ApplyIfNotNull(id => queryBuilder.AppendFormat("&id={0}", Uri.EscapeDataString(id)));

            if (string.IsNullOrEmpty(_transformer) == false)
                queryBuilder.AppendFormat("&transformer={0}", _transformer);

            return new GetRequest
            {
                Url = "/docs",
                Query = queryBuilder.ToString()
            };
        }

        public object Result { get; set; }
        public QueryResult QueryResult { get; set; }
        public bool RequiresRetry { get; set; }

        public void HandleResponse(GetResponse response)
        {
            if (response.ForceRetry)
            {
                Result = null;
                RequiresRetry = true;
                return;
            }

            var multiLoadResult = JsonDeserializationClient.GetDocumentResult((BlittableJsonReaderObject)response.Result);
            HandleResponse(multiLoadResult);
        }

        private void HandleResponse(GetDocumentResult loadResult)
        {
            _loadOperation.SetResult(loadResult);
            if (RequiresRetry == false)
                Result = _loadOperation.GetDocuments<T>();
        }
    }
}
